tensorflow/tsl/platform/strcat.cc

```diff
@@ -80,6 +80,9 @@ static char *Append4(char *out, const AlphaNum &x1, const AlphaNum &x2,
   memcpy(out, x3.data(), x3.size());
   out += x3.size();
 
+  if (x4.data() == nullptr)
+    return out;
+
```

Comment:
I'm not sure I understand why we check for x4 and none of the others? 

Also can you make sure that the unit-test for strcat covers this case? (and others here)

Reply:
@joker-eph The use case that triggered the undefined behaviour warning had the fourth item possibly be null. The first 3 items were static strings. This change results in the desired behaviour in that case. I am not sure what the correct behaviour would be if one of the earlier items was null.
Possible ways I saw to fix the reported problem

1. This suggested change
2. Create Append3 function and change call site to test for null and choose appropriate function Append3 or Append4, but that changes the interface and I am not sure what that would entail.
3. Change call site to check for the fourth item being null and modify it to be an empty string instead.
4. Put null check on all items in all versions of Append but then what to do if null found? Just skip item or return early or??

I chose 1. as being simplest but will modify it if an alternative is more desirable. Please tell me what you would like to see.

Reply:
I don't quite follow. You found a case where this method was called with a nullptr and discovered it does not null check before calling memcpy.


From there we have two possibilities:

1) The contract of all these helpers in strcat.cc is that passing in nullptr is UB and it is a bug on the caller side.

2) These helpers should be resilient to nullptr inputs and check to avoid UB.


If we are in case 2, then your option 1 does not seems like a valid way forward to me: you're fixing one example instead of the principled approach of fixing the contract of these API generally. That is not only Append4 but all of the helpers in this file.
As you mentioned that requires to specify that we would skip over an element if null (but that seems fine to me).


(and we should have tests still)

Reply:
@joker-eph  Fixed all Append helpers and added tests.